app/src/main/java/io/github/nicolasraoul/rosette/LanguageManager.kt

```diff
@@ -0,0 +1,123 @@
+package io.github.nicolasraoul.rosette
+
+import android.content.Context
+import android.content.SharedPreferences
+import android.util.Log
+import kotlinx.coroutines.CancellationException
+
+class LanguageManager(private val context: Context) {
+    private val preferences: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
+    private val wikipediaApiService = RetrofitClient.wikipediaApiService
+
+    companion object {
+        private const val TAG = "LanguageManager"
+        private const val PREFS_NAME = "language_preferences"
+        private const val KEY_DISPLAY_LANGUAGES = "display_languages"
+        private const val KEY_SEARCH_PRIORITY_LANGUAGES = "search_priority_languages"
+        
+        // Default languages - same as original hardcoded values
+        private val DEFAULT_DISPLAY_LANGUAGES = arrayOf("en", "fr", "ja")
+        private val DEFAULT_SEARCH_PRIORITY_LANGUAGES = arrayOf("fr", "ja", "en")
+    }
+
+    /**
+     * Get the current display languages (for the 3 WebViews)
+     */
+    fun getDisplayLanguages(): Array<String> {
+        val savedLanguages = preferences.getString(KEY_DISPLAY_LANGUAGES, null)
+        return if (savedLanguages != null) {
+            savedLanguages.split(",").toTypedArray()
+        } else {
+            DEFAULT_DISPLAY_LANGUAGES
+        }
+    }
+
+    /**
+     * Get the current search priority languages (order to search for articles)
+     */
+    fun getSearchPriorityLanguages(): Array<String> {
+        val savedLanguages = preferences.getString(KEY_SEARCH_PRIORITY_LANGUAGES, null)
+        return if (savedLanguages != null) {
+            savedLanguages.split(",").toTypedArray()
+        } else {
+            DEFAULT_SEARCH_PRIORITY_LANGUAGES
+        }
+    }
+
+    /**
+     * Save selected languages
+     */
+    fun saveLanguages(displayLanguages: Array<String>) {
+        // For simplicity, use the same order for both display and search priority
+        // This can be enhanced later if users need different search order
+        preferences.edit()
+            .putString(KEY_DISPLAY_LANGUAGES, displayLanguages.joinToString(","))
+            .putString(KEY_SEARCH_PRIORITY_LANGUAGES, displayLanguages.joinToString(","))
+            .apply()
+        Log.d(TAG, "Saved languages: ${displayLanguages.joinToString(", ")}")
+    }
+
+    /**
+     * Get list of available Wikipedia languages from the API
+     */
+    suspend fun getAvailableWikipediaLanguages(): List<WikipediaLanguage> {
+        try {
+            val apiUrl = "https://en.wikipedia.org/w/api.php?action=sitematrix&format=json&smtype=language"
```

Comment:
No, make the API service really output the actual URL it calls. Because I suspect it may not be the same one.